uast/uast-encoding-v2.md

```diff
@@ -0,0 +1,249 @@
+followed by `N` bytes of the message itself.
+
+There are only two messages used to encode the UAST:
+
+- [GraphHeader](https://github.com/bblfsh/sdk/blob/b44cc02214bf0d9c55dc3b830f423a6df6709ca9/uast/nodes/nodesproto/nodes.proto#L17)
+- [Node](https://github.com/bblfsh/sdk/blob/b44cc02214bf0d9c55dc3b830f423a6df6709ca9/uast/nodes/nodesproto/nodes.proto#L46)
+
+See the sections below for the usage of those messages.
+
+### Header
+
+The binary UAST file begins with a `\x00bgr` (`0x00 0x62 0x67 0x72`) "magic" that can be
+used to detect a format of the file.
+
+It is followed by a version number encoded as an unsigned 32 bit integer in little-endian byte order.
+The current format version is `1`.
+
+Next, a length-delimited `GraphHeader` message follows:
+```proto
+message GraphHeader {
+    uint64 last_id  = 1;
+    uint64 root     = 2;
+    uint64 metadata = 3;
+}
+```
+
+`last_id` is an optional field that sets a maximal node ID reserved by the node ID allocator.
+If not set, `max(node.ID)` is assumed. Implementation may reserve more IDs than required by
+the UAST nodes by setting `last_id > max(node.ID)`.
+
+`root` is an ID of the tree root node. Multiple unnamed roots can be stored by referencing an
+[Array](#array) node by its ID, and multiple named roots can be stored by referencing an
+[Object](#object) node. If not set explicitly, implementations should search the graph for
+unreferenced nodes of type [Array](#array) or [Object](#object) and treat them as an array of roots.
+
+`metadata` is an optional ID of a file metadata root, similar to `root`. This tree may
+contain arbitrary tree structure that describes the UAST file. If not set, no metadata is
+assumed. Implementation may ignore this tree, except for a case when `root` is not explicitly.
+
+### Nodes list
+
+Node list follows the [Header](#header) message. The list contains zero or more length-delimited
+[Node](https://github.com/bblfsh/sdk/blob/b44cc02214bf0d9c55dc3b830f423a6df6709ca9/uast/nodes/nodesproto/nodes.proto#L46)
+messages until the EOF. There is no explicit node count provided by the format.
+
+### Node
+
+Each Node is described by the following protocol buffer message:
+```proto
+message Node {
+    uint64 id = 1;
+    
+    oneof value {
+        string string = 2;
+        int64  int    = 3;
+        uint64 uint   = 4;
+        double float  = 5;
+        bool   bool   = 6;
+    }
+    
+    repeated uint64 keys = 7;
+    uint64 keys_from = 10;
+    
+    repeated uint64 values = 8;
+    
+    bool is_object = 9;
+    
+    uint64 values_offs = 11;
+}
+```
+
+The `id` field is unique, monotonically increasing value. ID value is used in most message
+fields to reference other nodes.
+
+If `id` field is omitted, the `prevNode.ID + 1` is assumed. The first node with no `id`
+is assumed to have an ID value of `1`.
+
+Nil nodes must not be represented by the Node message. Instead, they are represented by
```

Comment:
(For separate discussion) I wonder if we shouldn't reserve a chunk of the ID space for well-known objects like the empty node (is that what "nil" is meant to be?) Maybe start real nodes at 1000 and keep 1..999 for plumbing?

Reply:
First, as a clarification, the `nil` node is different from an empty object (`{}`) and an empty array (`[]`). Maybe this distinction is not important, but I wanted to make JSON->UAST->JSON transformation lossless. This helps make reversibility easier.

Answering the question, I considered to include the "static table" that reserves some ID space. But then decided not to do so, because the optimal table may be different for Native/Semantic ASTs.

For example, some drivers won't emit empty object/arrays at all. Since the encoded ID length varies depending on the value (`unsigned varint`), having lower IDs helps with the overall file size. Following the example, implementations that don't use empty arrays may benefit from lower IDs by reusing the reserved space.

At the same time, possible 3rd-party implementations might not care about the file size and may emit a table in any order they prefer (e.g. DFS).

So the table order depends on whatever the encoder cares about space efficiency or not. If it does, it may emit the best possible table for a specific AST, while still conforming with the spec. If it doesn't care, it may do whatever it wants with it.

The general intention behind this format was that it should be vendor-agnostic, so it's not declaring any hardcode any Babelfish-specific predicates (even the `@type`). It may be useful in the long term. If the format survives, of course :grinning:

Reply:
Re: Nil vs. empty. Thank you for the clarification, I didn't realize that distinction was important. The rationale makes sense. We might call it out:

> **Note:** Nil is a special node distinct from an empty object or array, used to distinguish a pointer to an absent node from an absent pointer to a node.

or something along those lines.

I can see the motivation for allowing variation—e.g., in the order of outputs. But I think we need to be prepared to put at least _some_ constraints on the emitter: If we don't even have basic structural constraints, it will be hard to use the graph downstream without baking in all the possible variations and their meaning. I do think it makes sense to avoid constructs that might be hard to implement, though.

As a rule of thumb, we should try not to constrain too much about what the driver chooses to put into the nodes, but we should be clear about the basic topological data. This is not something we have to resolve here and now, though.